Kaptain 2.0.0 Docs operations folder (#4317)
mesosphere/dcos-docs-site: pages/dkp/kaptain/2.0.0/operations/mysql-cluster/index.md

---
layout: layout.pug
navigationTitle: MySQL Cluster
title: Scalability and operations for the Kaptain MySQL Cluster
menuWeight: 70
excerpt: Scalability and operations for the Kaptain MySQL Cluster
beta: false
enterprise: false
---

Learn how to scale and backup/restore the MySQL cluster with Kaptain. Kaptain provides a consolidated
MySQL cluster with primary-primary replication for storage of the Pipelines execution history and artifacts,
and Katib experiment results. 

## Prerequisites

-   You already provisioned a Konvoy cluster using at least `v1.7.0`.

## Overview
By default, the MySQL cluster is based on Percona XtraDB, and consists of 3 database nodes and 2 proxy nodes. The database nodes run MySQL daemons and store the data, and the proxy nodes run HAProxy and expose the database to the clients via a Kubernetes Service.

> The database nodes form a primary-primary database cluster.  This means there are no secondary nodes; i.e. all data is consistent across all the cluster nodes whenever a database operation completes successfully.

Author: ```suggestion
The database nodes form a primary-primary database cluster. This means there are no secondary nodes; That is all data is consistent across all the cluster nodes whenever a database operation completes successfully.
```

> Horizontally scaling the database nodes will result in more replicas of the data available for recovery in case of a failure.  However, it will not result in better performance because of the replication model.

Author: ```suggestion
Horizontally scaling the database nodes will result in more replicas of the data available for recovery in case of a failure. However, it will not result in better performance because of the replication model.
```

To achieve better performance, the database nodes should be scaled vertically by adding more resources, such as CPU and memory.

> <p class="message--warning"><strong>WARNING: </strong>The Kaptain MySQL cluster does not support simultaneous changes to the database resources and the number of database nodes at the same time. If changes in both are required, you should perform them sequentially (i.e. increase resources, and then scale, or vice versa.)</p>

Author: ```suggestion
<p class="message--warning"><strong>WARNING: </strong>The Kaptain MySQL cluster does not support simultaneous changes to the database resources and the number of database nodes at the same time. If changes in both are required, you should perform them sequentially (that is increase resources, and then scale, or vice versa.)</p>
```

```bash
kubectl kudo update \
		--instance kaptain \
		-P parameters.yaml \
		--namespace kubeflow
```

### Scaling MySQL nodes
To change the number of the database nodes, create or update a configuration file named `parameters.yaml` and include the following property:
```yaml
datastoreNodeCount: "5"
```

The `datastoreNodeCount` property determines the number of MySQL instances in the cluster. The higher the number, the more copies of the data that will be stored. The default number of nodes is set to 3 for the best redundancy/throughput ratio. It is recommended to increase the number of replicas if you have unstable physical infrastructure or there is a high risk of data loss because of the unstable storage layers.

To install Kaptain with the provided parameters, run the following command on the cluster:
```bash
kubectl kudo install ./kubeflow-1.4.0_1.3.0.tgz \
		--instance kaptain \
		-P parameters.yaml \
		--namespace kubeflow \
		--create-namespace
```

To update an existing Kaptain instance with the provided parameters, run the following command on the cluster:
```bash
kubectl kudo update \
		--instance kaptain \
		-P parameters.yaml \
		--namespace kubeflow
```

> <p class="message--warning"><strong>WARNING: </strong>When scaling the existing database cluster down (i.e. decreasing the number of nodes), it is possible for clients to lose connectivity if the leader node is decommissioned. Once a new leader is elected, the clients will be able to reconnect.</p>

Author: ```suggestion
<p class="message--warning"><strong>WARNING: </strong>When scaling the existing database cluster down (that is decreasing the number of nodes), it is possible for clients to lose connectivity if the leader node is decommissioned. Once a new leader is elected, the clients will be able to reconnect.</p>
```

		-P parameters.yaml \
		--namespace kubeflow \
		--create-namespace
```

To update an existing Kaptain instance with the provided properties, run the following command on the cluster:
```bash
kubectl kudo update \
		--instance kaptain \
		-P parameters.yaml \
		--namespace kubeflow
```

To perform a backup, trigger the Kaptain `backup` plan for the MySQL cluster component:
```bash
kubectl kudo plan trigger --name=backup \
        --instance kaptain-percona-xtradb-cluster-operator \
        --namespace kubeflow
```

After the backup completes, it will be uploaded to the configured S3 bucket, for example:
```bash
aws s3 ls s3://kaptain-backup/
  28KiB kaptain-mysql-store-2021-04-05-23:49:57-full.md5
     0B kaptain-mysql-store-2021-04-05-23:49:57-full.sst_info/
     0B kaptain-mysql-store-2021-04-05-23:49:57-full/
```
## Restoring the MySQL cluster from a backup

<p class="message--warning"><strong>WARNING: </strong>The restore operation will terminate the running MySQL cluster and delete all the existing data. The data from the backup will be used to bootstrap a new cluster. The restore operation introduces downtime for the duration of the restore process.</p>

> There are two options for restoring the database cluster from a backup:

Author: ```suggestion
Restoring the database cluster from a backup is possible in two ways:
```
